module/state_synchronization/execution_data_service.go

```diff
@@ -0,0 +1,280 @@
+	"github.com/onflow/flow-go/network"
+)
+
+const (
+	defaultMaxBlobSize      = 1 << 20 // 1MiB
+	defaultMaxBlobTreeDepth = 2       // prevents malicious CID from causing download of unbounded amounts of data
+	defaultBlobBatchSize    = 16
+)
+
+var ErrBlobTreeDepthExceeded = errors.New("blob tree depth exceeded")
+
+// ExecutionDataService handles adding and getting execution data from a blobservice
+type ExecutionDataService struct {
+	serializer  *serializer
+	blobService network.BlobService
+	maxBlobSize int
+}
+
+func NewExecutionDataService(
+	codec encoding.Codec,
+	compressor network.Compressor,
+	blobService network.BlobService,
+) *ExecutionDataService {
+	return &ExecutionDataService{&serializer{codec, compressor}, blobService, defaultMaxBlobSize}
+}
+
+func (s *ExecutionDataService) receiveBatch(ctx context.Context, br *BlobReceiver) ([]network.Blob, error) {
+	var blobs []network.Blob
+	var err error
+
+	for i := 0; i < defaultBlobBatchSize; i++ {
+		var blob network.Blob
+		blob, err = br.Receive()
+
+		if err != nil {
+			break
+		}
+
+		blobs = append(blobs, blob)
+	}
+
+	return blobs, err
+}
+
+func (s *ExecutionDataService) receiveBlobs(parent context.Context, br *BlobReceiver) ([]cid.Cid, error) {
+	defer br.Close()
+
+	ctx, cancel := context.WithCancel(parent)
+	defer cancel()
+
+	var cids []cid.Cid
+
+	for {
+		batch, recvErr := s.receiveBatch(ctx, br)
+
+		for _, blob := range batch {
+			cids = append(cids, blob.Cid())
+		}
+
+		if err := s.blobService.AddBlobs(ctx, batch); err != nil {
+			return nil, err
+		}
+
+		if recvErr != nil {
+			if recvErr != ErrClosedBlobChannel {
+				// this is an unexpected error, and should never occur
+				return nil, recvErr
+			}
+
+			break
+		}
+	}
+
+	return cids, nil
+}
+
+func (s *ExecutionDataService) addBlobs(ctx context.Context, v interface{}) ([]cid.Cid, error) {
+	bcw, br := IncomingBlobChannel(s.maxBlobSize)
+
+	done := make(chan struct{})
+	var serializeErr error
+
+	go func() {
+		defer close(done)
+		defer bcw.Close()
+
+		if serializeErr = s.serializer.Serialize(bcw, v); serializeErr != nil {
+			return
+		}
+
+		serializeErr = bcw.Flush()
+	}()
+
+	cids, recvErr := s.receiveBlobs(ctx, br)
+
+	<-done
+
+	if recvErr != nil {
+		return nil, recvErr
+	}
+
+	return cids, serializeErr
+}
+
+// Add constructs a blob tree for the given ExecutionData and adds it to the blobservice, and then returns the root CID.
+func (s *ExecutionDataService) Add(ctx context.Context, sd *ExecutionData) (cid.Cid, error) {
+	cids, err := s.addBlobs(ctx, sd)
+
+	if err != nil {
+		return cid.Undef, fmt.Errorf("failed to add execution data blobs: %w", err)
+	}
+
+	for {
+		if len(cids) == 1 {
+			return cids[0], nil
+		}
+
+		if cids, err = s.addBlobs(ctx, cids); err != nil {
+			return cid.Undef, fmt.Errorf("failed to add cid blobs: %w", err)
+		}
+	}
+}
+
+func (s *ExecutionDataService) sendBlobs(parent context.Context, bs *BlobSender, cids []cid.Cid) error {
+	defer bs.Close()
```

Comment:
I don't see why this is better. It makes more sense to have it here because this function is the scope of the logic that actually uses bs.

Reply:
@Ullaakut So while I agree with most of what you're saying in most cases, the important thing to understand here is that we are not actually working with channels here. We are working with an abstraction which happens to be implemented using channels. As I've mentioned before, the implementation is based on the Go implementation of Pipe:
https://cs.opensource.google/go/go/+/refs/tags/go1.17.3:src/io/pipe.go

Instead of thinking about a BlobChannel as an actual channel, it would be more appropriate to think of it as a Pipe for blobs / blob data.